centaur/src/main/scala/centaur/json/JsonUtils.scala

```diff
@@ -26,25 +26,43 @@ object JsonUtils {
       * The second handles scatters, where each scatter has its own element in the array. In this case all of the
       * scatters are flattened themselves with the scatter index being inserted into the flattened key name.
       */
-    def flatten(prefix: String = ""): JsObject = jsValue.asJsObject.fields.foldLeft(JsObject.empty) {
-      case (acc, (k, v: JsArray)) if v.isSingleCallArray => acc ++ JsObject(k -> v.elements.head).flatten(prefix)
-      case (acc, (k, v: JsArray)) if v.hasShardIndex =>
-        // The .get on the shardIndex is safe as we know all elements of the array have a shardIndex field
-        acc ++ v.elements.map(_.asJsObject).fold(JsObject.empty) { (x, y) => x ++ y.flatten(s"$k.${y.shardIndex.get}") }
-      case (acc, (k, v: JsArray)) =>
-        v.elements.zipWithIndex.foldLeft(acc) { case (accumulator, (element, idx)) =>
-          val maybePrefix = if (prefix.isEmpty) "" else s"$prefix."
-          element match {
-            case _: JsObject => accumulator.mergeWith(element.flatten(s"$maybePrefix$k.$idx"))
-            case x: JsValue => accumulator + (s"$maybePrefix$k.$idx" -> x)
+    def flatten(prefix: String = ""): JsObject = {
+
+      def flattenShardAndAttempt(k: String, v: JsArray, f: JsObject => String): JsObject = {
+        v.elements.map(_.asJsObject).fold(JsObject.empty) { (x, y) => x ++ y.flatten(s"$k.${f(y)}") }
+      }
+
+      jsValue.asJsObject.fields.foldLeft(JsObject.empty) {
+        case (acc, (k, v: JsArray)) if v.isSingleCallArray => acc ++ JsObject(k -> v.elements.head).flatten(prefix)
+        case (acc, (k, v: JsArray)) if v.hasShardIndex && v.hasAttemptNumber =>
+          /* The .get on the shardIndex and attemptNumber is safe as we know all elements of the array have a
+             shardIndex and attempt field. This conversion will add the attempt number in the flattened key structure
+             to avoid lossy conversion for multiple attempts of the same shard
+          */
+          acc ++
+            flattenShardAndAttempt(k, v, (y: JsObject) => y.shardIndex.get) ++
+            flattenShardAndAttempt(k, v, (y: JsObject) => s"${y.shardIndex.get}.${y.attemptNumber.get}")
+        case (acc, (k, v: JsArray)) if v.hasShardIndex =>
+          /* This way of flattening shards is kept so that the new structure from above case doesn't fail all other
+              tests that rely on the older flattened structure. This should be cleaned up in
+              https://broadworkbench.atlassian.net/browse/BW-483
+           */
+          acc ++ flattenShardAndAttempt(k, v, (y: JsObject) => y.shardIndex.get)
+        case (acc, (k, v: JsArray)) =>
+          v.elements.zipWithIndex.foldLeft(acc) { case (accumulator, (element, idx)) =>
+            val maybePrefix = if (prefix.isEmpty) "" else s"$prefix."
+            element match {
```

Comment:
TOL this would be a breaking change but a really nice one IMO. I think we should replace this special-case `match` statement with a recursive call instead:
```scala
val maybePrefix = if (prefix.isEmpty) "" else "prefix."
val newPrefix = s"$maybePrefix$k.$idk"
element.flatten(newPrefix)
```

If I'm right, I think this is the reason why our `.test` expectations currently only unpack a single array index if there are nested arrays - and force you to type in the "result" array as a plaintext string (eg https://github.com/broadinstitute/cromwell/blob/develop/centaur/src/main/resources/standardTestCases/read_tsv.test#L17).

Reply:
NB: TOL and Optional!